Dockerfile

```diff
@@ -1,78 +1,138 @@
 
-# Adds cargo to path for rust dependencies
-ENV PATH                /root/.cargo/bin:$PATH
-
 # Path variables needed for Charm
 ENV LIBRARY_PATH        /usr/local/lib
 ENV LD_LIBRARY_PATH     /usr/local/lib
 
-# Downloads rust and sets it up
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain nightly-2018-10-24
+# Make sh point to bash
+# This is being changed since it will avoid any errors in the `launch_mpc.sh` script
+# which relies on certain code that doesn't work in container's default shell.
+RUN ln -sf bash /bin/sh
 
 # Install apt dependencies
-RUN apt-get update && apt-get install -y \
-    vim \
-    tmux \
+RUN apt-get update && apt-get install -y --no-install-recommends \
     bison \
+    curl \
     flex \
+    g++ \
+    git \
+    libflint-dev \
     libgmp-dev \
     libmpc-dev \
     libmpfr-dev \
-    libflint-dev
+    libssl-dev \
+    make \
+    openssl \
+    tmux \
+    wget \
+    vim 
+
+# This is needed otherwise the build for the power sum solver will fail.
+# This is a known issue in the version of libflint-dev in apt.
+# https://github.com/wbhart/flint2/issues/217
+# This has been fixed if we pull the latest code from the repo. However, we want
+# to avoid compiling the lib from the source since it adds 20 minutes to the build.
+RUN sed -i '30c #include "flint/flint.h"' /usr/include/flint/flintxx/flint_classes.h
+RUN echo "alias cls=\"clear && printf '\e[3J'\"" >> ~/.bashrc
+
+# RUN python -m venv /opt/venv
+# ENV PATH "/opt/venv/bin:${PATH}"
+
+# Downloads rust and sets it up
+RUN curl https://sh.rustup.rs -sSf | sh -s -- -y --default-toolchain nightly-2018-10-24
+ENV PATH "/root/.cargo/bin:${PATH}"
+
 # Download and build NTL from source
 # Shoup recommends not using O3
-RUN wget -qO- https://www.shoup.net/ntl/ntl-11.3.2.tar.gz | tar xzvf - \
-         &&  cd ntl-11.3.2/src \
-         && ./configure CXXFLAGS="-g -O2 -fPIC -march=native -pthread -std=c++11" \
-         && make -j4 \
-         && make install
+RUN curl -so - https://www.shoup.net/ntl/ntl-11.3.2.tar.gz | tar xzvf -
+WORKDIR /ntl-11.3.2/src  
+RUN ./configure CXXFLAGS="-g -O2 -fPIC -march=native -pthread -std=c++11" 
+RUN make -j 
+RUN make install -j
+WORKDIR /
 
-# Sets default directory for running the rest of the commands
-WORKDIR /usr/src/HoneyBadgerMPC
 
-RUN pip install --upgrade pip
+# Install betterpairing
+RUN curl -so - https://crypto.stanford.edu/pbc/files/pbc-0.5.14.tar.gz | tar xzvf - 
+WORKDIR /pbc-0.5.14/
+RUN ./configure
+RUN make -j
+RUN make install -j
+WORKDIR /
+
 
+# Downloads and installs charm
+RUN git clone https://github.com/JHUISI/charm.git 
+WORKDIR /charm/
+RUN git reset --hard be9587ccdd4d61c591fb50728ebf2a4690a2064f
+RUN ./configure.sh
+RUN make install -j
+WORKDIR /
+
+
+# Below derived from https://pythonspeed.com/articles/multi-stage-docker-python/
+WORKDIR /usr/src/HoneyBadgerMPC
 COPY . /usr/src/HoneyBadgerMPC
 
+RUN pip install --upgrade pip
+RUN pip install Cython
+RUN pip install -e .
```

Comment:
While you are at it, can you look into `requirements.txt` or some other way in which we can skip downloading the packages for every call to `docker-compose build`?

You can take it up in a separate PR if you want.

Reply:
Sure-- I think it'll be better.